src/riddley/walk.clj

```diff
@@ -206,71 +250,70 @@
-
-                  (instance? java.util.Map$Entry x)
-                  (clojure.lang.MapEntry.
-                    (walk-exprs' (key x))
-                    (walk-exprs' (val x)))
-
-                  (or
-                    (set? x)
-                    (vector? x))
-                  (into (empty x) (map walk-exprs' x))
-
-                  (instance? clojure.lang.IRecord x)
-                  x
-
-                  (map? x)
-                  (into (empty x) (map walk-exprs' x))
-
-                  ;; special case to handle clojure.test
-                  (and (symbol? x) (-> x meta :test))
-                  (vary-meta x update-in [:test] walk-exprs')
-
-                  :else
-                  x)]
-         (if (instance? clojure.lang.IObj x')
-           (with-meta x' (merge (meta x) (meta x')))
-           x')))))
+   (cmp/with-base-env
+     (let [x (try
+               (macroexpand x special-form?)
+               (catch ClassNotFoundException _
+                 x))
+           walk-exprs' (partial walk-exprs predicate handler special-form?)
+           x' (cond
+
+                (and (head= x 'var) (predicate x))
+                (handler (eval x))
+
+                (and (head= x 'quote)  (not (predicate x)))
+                x
+
+                (predicate x)
+                (handler x)
+
+                (seq? x)
+                (if (or (and (not try-clause?)
+                             (#{'catch 'finally} (first x)))
+                        (not (contains? special-forms (first x))))
+                  (doall (map walk-exprs' x))
+                  ((condp = (first x)
+                     'do        do-handler
+                     'def       def-handler
+                     'fn*       fn-handler
+                     'let*      let-handler
+                     'loop*     let-handler
+                     'letfn*    letfn-handler
+                     'case*     case-handler
+                     'try       try-handler
+                     'catch     catch-handler
+                     'reify*    reify-handler
+                     'deftype*  deftype-handler
+                     '.         dot-handler
+                     #(doall (map %1 %2)))
+                   walk-exprs' (special-meta x)))
+
+                (instance? java.util.Map$Entry x)
+                (clojure.lang.MapEntry.
+                 (walk-exprs' (key x))
+                 (walk-exprs' (val x)))
+
+                (or
+                 (set? x)
+                 (vector? x))
+                (into (empty x) (map walk-exprs' x))
+
+                (instance? clojure.lang.IRecord x)
+                x
+
+                (map? x)
+                (into (empty x) (map walk-exprs' x))
+
+                ;; special case to handle clojure.test
+                (and (symbol? x) (-> x meta :test))
+                (vary-meta x update-in [:test] walk-exprs')
+
+                :else
+                x)]
+       (merge-meta x' (meta x))))))
```

Comment:
The only change here was to fix the indentation and use the `merge-meta` helper function I wrote to simplify things a bit